com.unity.multiplayer.mlapi/Runtime/Core/MessageBatcher.cs

```diff
@@ -0,0 +1,198 @@
+        private Dictionary<ulong, SendStream> SendDict = new Dictionary<ulong, SendStream>();
+        private PooledBitWriter Writer = new PooledBitWriter(new PooledBitStream());
+
+        // Used to store targets, internally
+        private List<ulong> TargetList = new List<ulong>();
+
+        public void PushLength(int length, ref PooledBitWriter writer)
+        {
+            // If length is single byte we write it
+            if (length < 256)
+            {
+                writer.WriteByte((byte)length); // write the amounts of bytes that are coming up
+            }
+            else
+            {
+                // otherwise we write a 0, and a two-byte length
+                writer.WriteByte(0); // mark larger size
+                writer.WriteByte((byte)(length % 256)); // write the length modulo 256
+                writer.WriteByte((byte)(length / 256)); // write the length divided by 256
+            }
+        }
+
+        public int PopLength(in BitStream messageStream)
+        {
+            byte read = (byte)messageStream.ReadByte();
+            // if we read a non-zero value, we have a single byte length
+            if (read != 0)
+            {
+                return read;
+            }
+            // otherwise, a two-byte length follows
+            int length = messageStream.ReadByte();
+            length += messageStream.ReadByte() * 256;
+
+            return length;
+        }
+
+        /// <summary>
+        /// FillTargetList
+        /// Fills a list with the ClientId's an item is targeted to
+        /// </summary>
+        /// <param name="item">the FrameQueueItem we want targets for</param>
+        /// <param name="list">the list to fill</param>
+        private static void FillTargetList(in FrameQueueItem item, ref List<ulong> list)
+        {
+            list.Clear();
+            switch (item.QueueItemType)
+            {
+                case RPCQueueManager.QueueItemType.ServerRpc:
+                    list.Add(item.NetworkId);
+                    break;
+                case RPCQueueManager.QueueItemType.ClientRpc:
+                    foreach (ulong clientId in item.ClientIds)
+                    {
+                        list.Add(clientId);
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        /// <summary>
+        /// QueueItem
+        /// Add a FrameQueueItem to be sent
+        /// </summary>queueItem
+        /// <param name="item">the threshold in bytes</param>
+        public void QueueItem(in FrameQueueItem item)
+        {
+            FillTargetList(item, ref TargetList);
+
+            foreach (ulong clientId in TargetList)
+            {
+                // todo: actually queue and buffer. For now, the dict contains just one entry !!!
+
+                if (!SendDict.ContainsKey(clientId))
+                {
+                    SendDict[clientId] = new SendStream();
```

Comment:
Does this happen enough where we'd want a Pool?

Reply:
Definitely happens enough. 

What if instead of `SendDict.Remove(clientid);` I was resetting the SendStream Stream member, but leaving the SendStream entry in ? That way, once the SendDict contains an entry for a client (each new client) there's no further SendStream allocations ?

If this is sensible to you, I can start on it.

Reply:
Actually, keeping the SendStream in the Dict has awesome impact all-around. Much less allocation, removes the need for a second pass clean-up of the sent stuff, ... Much cleaner. Will update with the changes

Reply:
nice!